Fixing ng-validation states

jayellsbrg opened the pull request:
I wanted a way to add visual feedback, based on the state of the inputs. But couldn't rely on the existing states (e.g. $(un)touched, $pristine, & $dirty) since $dirty is set to true from the start, and $touched is set to true as soon as you open the pickadate menu.

This is the solution I propose.

Your thoughts on this?

jayellsbrg committed:
Add pickADate status indicators to ngModel

Adding pickADate status indicators to ngModel, since pickADate doesn't work well with ngModel's existing status indicators. Useful for visual feedback.

E.g.)
Once pickADate has completed initializing, ngModel's $dirty has already been set to true, and $pristine to false.
When the pickADate menu opens, $touched is set to true (and $untouched to false).

The new more accurate indicators are the four existing ($untouched, $touched, $pristine, $dirty), prefixed with pickADate:
    - $pickADateUntouched - Menu hasn't been (opened &) closed yet.
    - $pickADateTouched - Menu has been (opened &) closed.
    - $pickADatePristine - Date hasn't been changed.
    - $pickADateDirty - Date has been changed.

onSet seems to be called three times every time a pickADate is initialized, so $pickADateDirty isn't set to true until after the third call of onSet.

jayellsbrg committed:
Update README.md

Adding note for custom ngModel form control status indicators.

Toilal commented:
Thanks again for you contribution ! It seems fine from a pure code point of view, but i'll merge when i'm really sure it's not possible to rely on default angular form validation states, considering this as a workaround ...

I think it would be better to control $dirty and $touched from angular validation ... If ng-pickadate is included in a form that performs angular validation for other fields, ng-pickadate should not break the $dirty and $touched states of the whole form ...

Maybe it's possible to disable the automatic validation for this directive, and control it manually ?

jayellsbrg commented:
That's a good point about the validity of the whole form. I'll see if I can control the existing validation states.

jayellsbrg committed:
Revert "Update README.md"

This reverts commit b0ef8d8cf99fe40affd46b32de10a0cec252b274.

jayellsbrg committed:
Revert "Add pickADate status indicators to ngModel"

This reverts commit 6f0a8c7d04a1f303686a329bcd878679fdf3a665.

jayellsbrg committed:
Fix onClose arguments

jayellsbrg committed:
Fix onClose arguments

jayellsbrg committed:
Use ngModel to correct form validation issues

Fix ng-validation states

Fixing the following unexpected angular validation states caused by the pickadate jquery plugin:

- When pickadate sets the input text value, the input is marked as $dirty. -> To correct, set to $pristine at end of link function.
- When pickadate calendar opens, the input itself loses focus and is marked as $touched. -> To correct, set to $untouched the first time calendar opens, set to $touched when calendar closes.

Update README.md with ngModel validation info

Add note about how ng-pickadate manually controls angular form validation via ngModelController (when present).

Edit ngModel validation state section of README.md

Update README.md

jayellsbrg committed:
Fix ng-validation states

Fixing the following unexpected angular validation states caused by the pickadate jquery plugin:

- When pickadate sets the input text value, the input is marked as $dirty. -> To correct, set to $pristine at end of link function.
- When pickadate calendar opens, the input itself loses focus and is marked as $touched. -> To correct, set to $untouched the first time calendar opens, set to $touched when calendar closes.

jayellsbrg commented:
@Toilal Updated. Let me know what you think.

Toilal commented:
I've read the code and It seems to be really good. I'll give a test during the day, and probably merge this and release as `0.3.0`.

If you have some time, could you please :
- Update README.md with the ng-model changes ?
- Rebase everything in a single commit ?

I'll update the demo page just before releasing. Thanks a lot !

jayellsbrg committed:
Update README.md with ngModel validation info

Add note about how ng-pickadate manually controls angular form validation via ngModelController (when present).

jayellsbrg committed:
Edit ngModel validation state section of README.md

jayellsbrg committed:
Update README.md

jayellsbrg committed:
Merge branch 'master' of https://github.com/jayellsbrg/ng-pickadate

jayellsbrg commented:
Updated README.md and rebased commits into one.

jayellsbrg committed:
Merge pull request #1 from Toilal/master

Picking up release version bump from original

Toilal commented:
It seems you don't have rebased, the history is weird ... No problems, I review the whole and i'll rebase by myself ...

jayellsbrg commented:
Yeah, I did an interactive rebase and squashed all the commits into one. It
looked like it worked locally, but then for some reason on Github both the
rebase commit and the commits it should have included showed up. I had
never used rebase before.

If you can fix it, would you mind letting me know how you do it? I like
that you keep the commits clean like that. Being able to commit changes
progressively and then rebase them later into one large commit will come in
handy.

Thanks.

On Thu, Dec 3, 2015 at 9:45 AM, Rémi Alvergnat notifications@github.com
wrote:

> It seems you don't have rebased, the history is weird ... No problems, I
> review the whole and i'll rebase by myself ...
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Toilal/ng-pickadate/pull/32#issuecomment-161728201.

Toilal commented:
In this case, I didn't performed interactive rebase. I've created a new branch based on the head of your branch. And then, I've performed a "soft" reset to master. It brings the history to master, while keeping all changes inside the current index. Then I only have to perform a commit, and it's done :).

I use [smartgit](http://www.syntevo.com/smartgit/), it's really great. It supports everything but the interactive rebase, where I still need to use the command line.

jayellsbrg commented:
Cool. Thanks for sharing. :)

On Thu, Dec 3, 2015 at 10:16 AM, Rémi Alvergnat notifications@github.com
wrote:

> In this case, I didn't performed interactive rebase. I've created a new
> branch based on the head of your branch. And then, I've performed a "soft"
> reset to master. It brings the history to master, while keeping all changes
> inside the current index. Then I only have to perform a commit, and it's
> done :).
> 
> I use smartgit http://www.syntevo.com/smartgit/, it's really great. It
> supports everything but the interactive rebase, where I still need to use
> the command line.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Toilal/ng-pickadate/pull/32#issuecomment-161736706.

Toilal commented:
Also when your are creating a pull request, keep it mind that future changes in the branch will reflect into the pull request until it's merged. It seems you have released `0.3.0` on your branch.

Toilal commented:
So you have to create another branch for your own work you don't want to get merged in the origin repository.

Toilal commented:
It's now in master, thanks !

Toilal commented:
And released as `0.3.0`

jayellsbrg commented:
Great, thanks!

Sorry about that. I released mine as 0.3.0 so I could commit a working bower dependency in the meantime for other developers I'm working with. I'll create a separate branch for temporary releases in the future.

Toilal commented:
No problem, thanks again for contributing because I don't have much time
for this project and don't use it at the moment. I'll try to update the
website with latest demo too.
Le 3 déc. 2015 7:52 PM, "J Seaberg" notifications@github.com a écrit :

> Great, thanks!
> 
> Sorry about that. I released mine as 0.3.0 so I could commit a working
> bower dependency in the meantime for other developers I'm working with.
> I'll create a separate branch for temporary releases in the future.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/Toilal/ng-pickadate/pull/32#issuecomment-161745890.
